### PR TITLE
refactor: Convert Sound to Kotlin 

### PR DESCRIPTION
**Pull request by david-allison**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt [hunk 1]
```diff
@@ -15,146 +15,83 @@
...
-     * Pattern used to identify the markers for sound files
-     */
-    public static final Pattern SOUND_PATTERN = Pattern.compile("\\[sound:([^\\[\\]]*)]");
-
-    /**
-     * Pattern used to parse URI (according to http://tools.ietf.org/html/rfc3986#page-50)
-     */
-    private static final Pattern sUriPattern = Pattern.compile("^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?$");
-
+class Sound {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt:49]: I think we should add(here and for other big classes being migrated):

`@KotlinCleanup("improve kotlin code where ever possible")`

as an alternative to littering the class with a lot of smaller cleanups.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt [hunk 2]
```diff
@@ -15,146 +15,83 @@
...
+import android.net.Uri
+import android.os.Build
+import android.webkit.MimeTypeMap
+import android.widget.VideoView
+import androidx.annotation.CheckResult
+import androidx.annotation.VisibleForTesting
+import com.ichi2.anki.AbstractFlashcardViewer
+import com.ichi2.anki.AnkiDroidApp
+import com.ichi2.anki.ReadText.stopTts
+import com.ichi2.compat.CompatHelper.Companion.compat
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt:34]: style: keep the CompatHelper import as in the original code, using `compat.something()` is not that great.

Same requests for:
```
import com.ichi2.anki.ReadText.stopTts
import com.ichi2.utils.DisplayUtils.getDisplayDimensions
import net.ankiweb.rsdroid.BackendFactory.defaultLegacySchema
```

**Review by lukstbit — CHANGES_REQUESTED** [review]

File: AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt (hunk 1 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt:49]: I prefer the opposite - I use `@KotlinCleanup("lots to do")` to imply that the annotation should be broken down into more targeted annotations.

It's good to list acceptance criteria associated with an issue. Having a fixed number of easy cleanups is better than one general "complete everything" cleanup, as it's harder to determine when this cleanup annotation is completed.

File: AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt [hunk 3]
```diff
@@ -300,320 +175,425 @@ public class Sound {
...
-        }
-
-        return CompatHelper.getCompat().hasVideoThumbnail(path);
-    }
-
-
-    public String getCurrentAudioUri() {
-        if (mCurrentAudioUri == null) {
-            return null;
+            return false
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt:360]: @KotlinCleanup("simplify code with ?. ")

File: AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt [hunk 4]
```diff
@@ -300,320 +175,425 @@ public class Sound {
...
-                                .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
-                                .build());
-                mMediaPlayer.setOnPreparedListener(mp -> {
-                    Timber.d("Starting media player");
-                    mMediaPlayer.start();
-                });
-                mMediaPlayer.setOnCompletionListener(playAllListener);
-                mMediaPlayer.prepareAsync();
-                Timber.d("Requesting audio focus");
+                mMediaPlayer!!.setDataSource(
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Sound.kt:317]: @KotlinCleanup("simplify with scope function on mediaPlayer ")

**Review by lukstbit — CHANGES_REQUESTED** [review]
I've added some of the cleanups that I had in mind when first reviewing, everything looks ok.

**Review by lukstbit — APPROVED** [review]

**Comment by BrayanDSO** [comment]
This one will need a rebase into main to replace "AnkiDroidApp.getInstance()"

**Review by BrayanDSO — APPROVED** [review]

**Pull request merged**